sql/subquery.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+func (p *planner) expandSubqueries(expr parser.Expr) (parser.Expr, error) {
+	v := subqueryVisitor{planner: p}
+	expr = parser.WalkExpr(&v, expr)
+	return expr, v.err
+}
+
+type subqueryVisitor struct {
+	*planner
+	path []parser.Expr // parent expressions
+	err  error
+}
+
+var _ parser.Visitor = &subqueryVisitor{}
+
+func (v *subqueryVisitor) Visit(expr parser.Expr, pre bool) (parser.Visitor, parser.Expr) {
+	if v.err != nil {
+		return nil, expr
+	}
+	if !pre {
+		v.path = v.path[:len(v.path)-1]
+		return nil, expr
+	}
+	v.path = append(v.path, expr)
+
+	subquery, ok := expr.(*parser.Subquery)
+	if !ok {
+		return v, expr
+	}
+
+	var plan planNode
+	if plan, v.err = v.makePlan(subquery.Select); v.err != nil {
+		return nil, expr
+	}
+
+	columns, multipleRows := v.getSubqueryContext()
+	if columns != len(plan.Columns()) {
+		switch columns {
+		case 1:
+			v.err = fmt.Errorf("subquery must return only one column")
+		default:
+			v.err = fmt.Errorf("subquery must return %d columns", columns)
+		}
+		return nil, expr
+	}
+
+	var result parser.Expr
+	if multipleRows {
+		var rows parser.DTuple
+		for plan.Next() {
+			values := plan.Values()
+			switch len(values) {
+			case 1:
+				// This seems hokey, but if we don't do this then the subquery expands to
+				// a tuple of tuples instead of a tuple of values and an expression like
+				// "k IN (SELECT foo FROM bar)" will fail because we're comparing a
+				// single value against a tuple. Perhaps comparison of a single value
```

Comment:
I need to trim this comment. I think this is the only place this occurs, but please correct me if you can find another.